provider/datastore/keystore.go

```diff
@@ -0,0 +1,431 @@
+package datastore
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	query "github.com/ipfs/go-datastore/query"
+	"github.com/libp2p/go-libp2p-kad-dht/provider/internal/keyspace"
+	mh "github.com/multiformats/go-multihash"
+
+	"github.com/probe-lab/go-libdht/kad"
+	"github.com/probe-lab/go-libdht/kad/key/bit256"
+	"github.com/probe-lab/go-libdht/kad/key/bitstr"
+)
+
+// KeyStore indexes multihashes by their kademlia identifier.
+type KeyStore struct {
+	lk sync.Mutex
+
+	ds         ds.Batching
+	base       ds.Key
+	prefixBits int
+	batchSize  int
+}
+
+type keyStoreCfg struct {
+	base       string
+	prefixBits int
+	batchSize  int
+}
+
+// KeyStoreOption configures KeyStore behaviour.
+type KeyStoreOption func(*keyStoreCfg) error
+
+const (
+	DefaultKeyStoreBasePrefix = "/provider/keystore"
+	DefaultKeyStoreBatchSize  = 1 << 14
+	DefaultPrefixBits         = 16
+)
+
+var KeyStoreDefaultCfg = func(cfg *keyStoreCfg) error {
```

Comment:
It seems strange to have basically a option to set defaults, and looks a bit odd when options are being evaluated. Consider making this a function that just returns a `keyStoreCfg` with the members set to default values. Then apply the options to that.

Or, create a function that sets the defaults and then applies all the options, like [this](https://gist.github.com/gammazero/e453204dcdf808aed0bd2731b09779ad#file-option-go-L22-L35).

Reply:
Updated in https://github.com/libp2p/go-libp2p-kad-dht/pull/1150